File: src/installers/git_hook.ts
import { execSync } from "node:child_process";
import { existsSync } from "node:fs";
import { chmod, readFile } from "node:fs/promises";
import { join } from "node:path";
import { GIT_TEMPLATE_DIR, NOTES_REF } from "../config.ts";
import { writeAtomic } from "../util/atomic.ts";
import { gitBypassEnv } from "../util/git.ts";
import type { InstallResult } from "./base.ts";

const HOOK_MARKER = "# posthook v2";

function hookScript(binaryPath: string): string {
  // Self-healing notes transport: ensures fetch/push refspecs exist for refs/notes/posthook
  // on `origin` if they're missing. Means repos created via the global git template
  // also start syncing notes without needing `posthook track`. Safe to fail silently:
  // if `origin` doesn't exist yet, the config lines still get added and take effect once
  // `origin` is configured. No-op when the refspecs are already present.
  return `#!/bin/sh
${HOOK_MARKER}
# Captures commit metadata after every successful commit. Safe to fail silently.
"${binaryPath}" ingest --kind git-commit --repo-root "$(git rev-parse --show-toplevel)" --sha "$(git rev-parse HEAD)" >/dev/null 2>&1 || true
{
  spec='${NOTES_REF}:${NOTES_REF}'
  if ! git config --get-all remote.origin.fetch 2>/dev/null | grep -Fxq "$spec"; then
    git config --add remote.origin.fetch "$spec" 2>/dev/null || true
  fi
  if ! git config --get-all remote.origin.push 2>/dev/null | grep -Fxq "$spec"; then
    git config --add remote.origin.push "$spec" 2>/dev/null || true
  fi
} >/dev/null 2>&1 || true
`;
}

export function configureNotesTransport(repoPath: string): boolean {
  const spec = `${NOTES_REF}:${NOTES_REF}`;
  let changed = false;
  for (const key of ["remote.origin.fetch", "remote.origin.push"]) {
    let existing = "";
    try {
      existing = execSync(`git config --get-all ${key}`, {
        cwd: repoPath,
        encoding: "utf8",
        env: gitBypassEnv(),
      });
    } catch {
      // section missing — config --add will create it
    }
    if (!existing.split("\n").includes(spec)) {
      execSync(`git config --add ${key} ${JSON.stringify(spec)}`, {
        cwd: repoPath,
        env: gitBypassEnv(),
      });
      changed = true;
    }
  }
  return changed;
}

async function writeHookFile(path: string, binaryPath: string): Promise<boolean> {
  const desired = hookScript(binaryPath);
  if (existsSync(path)) {
    const existing = await readFile(path, "utf8");
    if (existing === desired) return false;
    if (!existing.includes(HOOK_MARKER) && existing.trim() !== "") {
      // Don't clobber a user-authored hook. Append a marker noting we skipped.
      throw new Error(
        `Refusing to overwrite existing post-commit hook at ${path}. Move it aside and rerun.`,
      );
    }
  }
  await writeAtomic(path, desired);
  await chmod(path, 0o755);
  return true;
}

export async function installGlobalGitTemplate(binaryPath: string): Promise<InstallResult> {
  const hooksDir = join(GIT_TEMPLATE_DIR, "hooks");
  const hookPath = join(hooksDir, "post-commit");
  const changed = await writeHookFile(hookPath, binaryPath);

  // Wire git to use this template for `git init` going forward.
  let configChanged = false;
  try {
    const current = execSync("git config --global --get init.templateDir", {
      encoding: "utf8",
      env: gitBypassEnv(),
    }).trim();
    if (current !== GIT_TEMPLATE_DIR) {
      execSync(`git config --global init.templateDir "${GIT_TEMPLATE_DIR}"`, {
        env: gitBypassEnv(),
      });
      configChanged = true;
    }
  } catch {
    execSync(`git config --global init.templateDir "${GIT_TEMPLATE_DIR}"`, {
      env: gitBypassEnv(),
    });
    configChanged = true;
  }

  const any = changed || configChanged;
  return {
    changed: any,
    path: hookPath,
    message: any
      ? `Git template: hook installed at ${hookPath}; init.templateDir set`
      : `Git template: already up to date`,
  };
}

export async function installRepoHook(repoPath: string, binaryPath: string): Promise<InstallResult> {
  const gitDir = execSync("git rev-parse --git-dir", {
    cwd: repoPath,
    encoding: "utf8",
    env: gitBypassEnv(),
  }).trim();
  const absGitDir = gitDir.startsWith("/") ? gitDir : join(repoPath, gitDir);
  const hookPath = join(absGitDir, "hooks", "post-commit");
  const hookChanged = await writeHookFile(hookPath, binaryPath);
  let notesChanged = false;
  try {
    notesChanged = configureNotesTransport(repoPath);
  } catch {
    // not fatal — the hook itself will retry the config on first commit
  }
  const changed = hookChanged || notesChanged;
  const parts: string[] = [];
  if (hookChanged) parts.push("installed");
  else parts.push("hook up to date");
  if (notesChanged) parts.push("notes transport configured");
  return {
    changed,
    path: hookPath,
    message: `Repo hook (${hookPath}): ${parts.join(", ")}`,
  };
}

File: internal/installers/githook.go
package installers

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/bilanc/posthook/internal/atomicfs"
	"github.com/bilanc/posthook/internal/gitx"
	"github.com/bilanc/posthook/internal/notes"
	"github.com/bilanc/posthook/internal/paths"
)

const hookMarker = "# posthook v3"

func hookScript(binaryPath string) string {
	// Fallback mode (no git shadow) can only observe commits, so the hook
	// configures the notes fetch refspec here: an ordinary `git fetch` then
	// brings teammates' notes down into the tracking ref, and `posthook
	// blame` reads that ref directly. Pushing notes in this mode is manual
	// (`posthook notes push`) — the shadow does it automatically.
	return "#!/bin/sh\n" +
		hookMarker + "\n" +
		"# Captures commit metadata after every successful commit. Safe to fail silently.\n" +
		`"` + binaryPath + `" ingest --kind git-commit --repo-root "$(git rev-parse --show-toplevel)" --sha "$(git rev-parse HEAD)" >/dev/null 2>&1 || true` + "\n" +
		`"` + binaryPath + `" notes configure "$(git rev-parse --show-toplevel)" >/dev/null 2>&1 || true` + "\n"
}

// ConfigureNotesTransport installs posthook's notes fetch refspec on origin
// and removes the legacy fetch/push refspecs. Returns true iff config was
// modified. See package notes for why fetch goes through a tracking ref.
func ConfigureNotesTransport(repoPath string) (bool, error) {
	return notes.EnsureRefspec(repoPath, "origin"), nil
}

func writeHookFile(path, binaryPath string) (bool, error) {
	desired := hookScript(binaryPath)
	if existing, err := os.ReadFile(path); err == nil {
		if string(existing) == desired {
			return false, nil
		}
		if !strings.Contains(string(existing), hookMarker) && strings.TrimSpace(string(existing)) != "" {
			return false, fmt.Errorf(
				"refusing to overwrite existing post-commit hook at %s. Move it aside and rerun.",
				path)
		}
	} else if !errors.Is(err, os.ErrNotExist) {
		return false, err
	}
	if err := atomicfs.WriteString(path, desired); err != nil {
		return false, err
	}
	if err := os.Chmod(path, 0o755); err != nil {
		return false, err
	}
	return true, nil
}

// InstallGlobalGitTemplate writes a post-commit hook into the global git
// template dir and points init.templateDir at it. New repos created via
// `git init` then auto-install the hook with no per-repo step.
func InstallGlobalGitTemplate(binaryPath string) (Result, error) {
	hooksDir := filepath.Join(paths.GitTemplateDir(), "hooks")
	hookPath := filepath.Join(hooksDir, "post-commit")
	changed, err := writeHookFile(hookPath, binaryPath)
	if err != nil {
		return Result{}, err
	}

	// Set init.templateDir if not already pointing here.
	current := gitx.Run("", "config", "--global", "--get", "init.templateDir")
	configChanged := false
	if current != paths.GitTemplateDir() {
		gitx.Run("", "config", "--global", "init.templateDir", paths.GitTemplateDir())
		configChanged = true
	}

	any := changed || configChanged
	msg := "Git template: already up to date"
	if any {
		msg = fmt.Sprintf("Git template: hook installed at %s; init.templateDir set", hookPath)
	}
	return Result{Changed: any, Path: hookPath, Message: msg}, nil
}

// InstallRepoHook installs the per-repo post-commit hook in repoPath. Used by
// `posthook track` for existing repos where the global template's auto-install
// won't kick in.
func InstallRepoHook(repoPath, binaryPath string) (Result, error) {
	gitDir := gitx.Run(repoPath, "rev-parse", "--git-dir")
	if gitDir == "" {
		return Result{}, fmt.Errorf("%s does not appear to be a git repo", repoPath)
	}
	absGitDir := gitDir
	if !filepath.IsAbs(absGitDir) {
		absGitDir = filepath.Join(repoPath, gitDir)
	}
	hookPath := filepath.Join(absGitDir, "hooks", "post-commit")
	hookChanged, err := writeHookFile(hookPath, binaryPath)
	if err != nil {
		return Result{}, err
	}
	notesChanged, _ := ConfigureNotesTransport(repoPath)

	parts := []string{}
	if hookChanged {
		parts = append(parts, "installed")
	} else {
		parts = append(parts, "hook up to date")
	}
	if notesChanged {
		parts = append(parts, "notes transport configured")
	}
	return Result{
		Changed: hookChanged || notesChanged,
		Path:    hookPath,
		Message: fmt.Sprintf("Repo hook (%s): %s", hookPath, strings.Join(parts, ", ")),
	}, nil
}
